openmc/material.py

```diff
@@ -0,0 +1,1109 @@
+        density = group['atom_density'][()]
+        if 'nuclide_densities' in group:
+            nuc_densities = group['nuclide_densities'][()]
+
+        # Create the Material
+        material = cls(mat_id, name)
+        material.depletable = bool(group.attrs['depletable'])
+        if 'volume' in group.attrs:
+            material.volume = group.attrs['volume']
+        if "temperature" in group.attrs:
+            material.temperature = group.attrs["temperature"]
+
+        # Read the names of the S(a,b) tables for this Material and add them
+        if 'sab_names' in group:
+            sab_tables = group['sab_names'][()]
+            for sab_table in sab_tables:
+                name = sab_table.decode()
+                material.add_s_alpha_beta(name)
+
+        # Set the Material's density to atom/b-cm as used by OpenMC
+        material.set_density(density=density, units='atom/b-cm')
+
+        if 'nuclides' in group:
+            nuclides = group['nuclides'][()]
+            # Add all nuclides to the Material
+            for fullname, density in zip(nuclides, nuc_densities):
+                name = fullname.decode().strip()
+                material.add_nuclide(name, percent=density, percent_type='ao')
+        if 'macroscopics' in group:
+            macroscopics = group['macroscopics'][()]
+            # Add all macroscopics to the Material
+            for fullname in macroscopics:
+                name = fullname.decode().strip()
+                material.add_macroscopic(name)
+
+        return material
+
+    def add_volume_information(self, volume_calc):
+        """Add volume information to a material.
+
+        Parameters
+        ----------
+        volume_calc : openmc.VolumeCalculation
+            Results from a stochastic volume calculation
+
+        """
+        if volume_calc.domain_type == 'material':
+            if self.id in volume_calc.volumes:
+                self._volume = volume_calc.volumes[self.id].n
+                self._atoms = volume_calc.atoms[self.id]
+            else:
+                raise ValueError('No volume information found for this material.')
+        else:
+            raise ValueError('No volume information found for this material.')
+
+    def set_density(self, units, density=None):
+        """Set the density of the material
+
+        Parameters
+        ----------
+        units : {'g/cm3', 'g/cc', 'kg/m3', 'atom/b-cm', 'atom/cm3', 'sum', 'macro'}
+            Physical units of density.
+        density : float, optional
+            Value of the density. Must be specified unless units is given as
+            'sum'.
+
+        """
+
+        cv.check_value('density units', units, DENSITY_UNITS)
+        self._density_units = units
+
+        if units == 'sum':
+            if density is not None:
+                msg = 'Density "{}" for Material ID="{}" is ignored ' \
+                      'because the unit is "sum"'.format(density, self.id)
+                warnings.warn(msg)
```

Comment:
this seems like programmer-error; the caller is expecting this to have side-effects -- I would document why you don't raise this msg